### PR TITLE
Make DOCKER_USERNAME and _PASSWORD optional

### LINKED ISSUE #70: Support pushing to multiple registries
**Is your feature request related to a problem? Please describe.**

When used with JupyterHubs, you often want to push the built image to two registries - a private one (like AWS ECR) closer to your JupyterHub, and a public one (like DockerHub or quay.io) for use with mybinder.org / others.

**Describe the solution you'd like**

Some way to specify additional tags the image should have, multiple forms of authentication, and an indicator that they should all be pushed.

**Describe alternatives you've considered**

This can be currently done with an extra step in the github workflow, doing a `docker tag` and a `docker push`.

### PR DESCRIPTION
**Pull request by consideRatio**
The idea is that by making these fields optional, we support letting the
user of the action pre-configure credentials or use already configured
credentials instead.

Closes #70

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -39,5 +39,5 @@ jobs:
       uses: ./
       with:
         NO_PUSH: true
+        DOCKER_USERNAME: ${{ github.actor }}
         MYBINDERORG_TAG: ${{ github.sha }}
-
```
Comment (consideRatio) [inline .github/workflows/test.yaml:43]: action.yml declared DOCKER_USERNAME as required, and this test didn't explicitly set DOCKER_USERNAME but instead relied on a script to populate the INPUT_DOCKER_USERNAME variable with the github actor.

I think it is cleaner to do it like this though. I was clueless about the scripts logic as it seemed like something that shouldn't be able to happen until I saw the GitHub workflow CI failure.

File: README.md [hunk 2]
```diff
@@ -48,27 +48,22 @@ Images generated by this action are automatically tagged with both `latest` and
...
 ## Optional Inputs
 
+- **`DOCKER_USERNAME`**:
+    description: Docker registry username.  If not supplied, credentials must be setup ahead of time.
+- **`DOCKER_PASSWORD`**:
+    description: Docker registry password or [access token (recommended)](https://docs.docker.com/docker-hub/access-tokens/).  If not supplied, credentials must be setup ahead of time.
+- **`DOCKER_REGISTRY`**:
+    description: name of the docker registry.  If not supplied, this defaults to [DockerHub](https://hub.docker.com/)
+- **`IMAGE_NAME`**:
+    name of the image.  Example - myusername/myContainer.  If not supplied, this defaults to `<DOCKER_USERNAME/GITHUB_REPOSITORY_NAME>`.
```
Comment (consideRatio) [inline README.md:60]: I moved DOCKER_USERNAME and PASSWORD from the mandatory section to the optional section. The other inputs were relocated to have the same order as in the action.yml file.

File: create_docker_image.sh [hunk 3]
```diff
@@ -18,22 +18,20 @@ if [ "$INPUT_APPENDIX_FILE" ]; then
...
     echo ${INPUT_DOCKER_PASSWORD} | docker login $INPUT_DOCKER_REGISTRY -u ${INPUT_DOCKER_USERNAME} --password-stdin
 fi
 
 REPO_NAME=`echo $GITHUB_REPOSITORY | cut -d "/" -f 2`
 
-# Set Docker username to the actor name not provided
-if [ -z "$INPUT_DOCKER_USERNAME" ]; then
-    INPUT_DOCKER_USERNAME="$GITHUB_ACTOR"
-fi
-
```
Comment (consideRatio) [inline create_docker_image.sh:34]: This logic was only applied during a CI test as DOCKER_USERNAME was declared required in action.yml. I removed this section in favor of making the CI test explicitly set DOCKER_USERNAME to ${{ github.actor }} instead as discussed in a comment above.

**Review by consideRatio — COMMENTED** [review]
I wrote some notes as all the changes are bundled to a single commit. Ping @yuvipanda who opened #70 that this PR is meant to close.

**Comment by hamelsmu** [comment]
LGTM

File: README.md [hunk 4]
```diff
@@ -48,27 +48,22 @@ Images generated by this action are automatically tagged with both `latest` and
...
-    description: Docker registry password or [access token](https://docs.docker.com/docker-hub/access-tokens/).  **If using DockerHub, we recommend using an access token instead of your password.**
-
 ## Optional Inputs
 
+- **`DOCKER_USERNAME`**:
+    description: Docker registry username.  If not supplied, credentials must be setup ahead of time.
+- **`DOCKER_PASSWORD`**:
+    description: Docker registry password or [access token (recommended)](https://docs.docker.com/docker-hub/access-tokens/).  If not supplied, credentials must be setup ahead of time.
+- **`DOCKER_REGISTRY`**:
+    description: name of the docker registry.  If not supplied, this defaults to [DockerHub](https://hub.docker.com/)
```
Comment (manics) [inline README.md:58]:
```suggestion
    description: domain name of the docker registry.  If not supplied, this defaults to [DockerHub](https://hub.docker.com/)
```
Or should it be a URL?

File: README.md [hunk 5]
```diff
@@ -282,6 +279,29 @@ jobs:
...
+    - name: checkout files in repo
+      uses: actions/checkout@main
+
+    # TODO: add a step here to setup credentials to push to your
+    #       docker registry before running the repo2docker-action
+
+    - name: update jupyter dependencies with repo2docker
+      uses: jupyterhub/repo2docker-action@master
+      with:
+        DOCKER_REGISTRY: your-registry
```
Comment (manics) [inline README.md:301]: Any chance you could make `your-registry` a full example, e.g. is it `your-registry.example.org` or `https://your-registry.example.org`?

**Review by manics — COMMENTED** [review]
I've made a couple of comments on improving the `DOCKER_REGISTRY` description.

File: README.md (hunk 5 above)
Comment (consideRatio) [reply README.md:301]: It should be `your-registry.example.org`!

File: action.yml [hunk 6]
```diff
@@ -2,11 +2,11 @@ name: 'repo2docker Action'
...
-    require: true
+    description: Docker registry username.  If not supplied, credentials must be setup ahead of time.
+    require: false
   DOCKER_PASSWORD:
-    description: Docker registry password
-    required: true
+    description: Docker registry password.  If not supplied, credentials must be setup ahead of time.
+    required: false
   DOCKER_REGISTRY:
     description: name of the docker registry.  If not supplied, this defaults to registry.hub.docker.com.
```
Comment (consideRatio) [inline action.yml:11]:
```suggestion
    description: domain name of the docker registry.  If not supplied, this defaults to registry.hub.docker.com.
```

File: README.md (hunk 5 above)
Comment (consideRatio) [reply README.md:301]:
```suggestion
        DOCKER_REGISTRY: your-registry.example.org
```

File: README.md [hunk 7]
```diff
@@ -48,27 +48,22 @@ Images generated by this action are automatically tagged with both `latest` and
...
 ## Optional Inputs
 
+- **`DOCKER_USERNAME`**:
+    description: Docker registry username.  If not supplied, credentials must be setup ahead of time.
+- **`DOCKER_PASSWORD`**:
+    description: Docker registry password or [access token (recommended)](https://docs.docker.com/docker-hub/access-tokens/).  If not supplied, credentials must be setup ahead of time.
+- **`DOCKER_REGISTRY`**:
+    description: domain name of the docker registry.  If not supplied, this defaults to [DockerHub](https://hub.docker.com/)
+- **`IMAGE_NAME`**:
+    name of the image.  Example - myusername/myContainer.  If not supplied, this defaults to `<DOCKER_USERNAME/GITHUB_REPOSITORY_NAME>`.
```
Comment (yuvipanda) [inline README.md:60]:
```suggestion
    name of the image.  Example - myusername/myContainer.  If not supplied, this defaults to `<DOCKER_USERNAME>/<GITHUB_REPOSITORY_NAME>`.
```

Comment (yuvipanda) [reply README.md:60]: To clarify that the `/` is part of the string, not the substitution

**Comment by yuvipanda** [comment]
Thanks a lot, @consideRatio and @manics!!!

File: create_docker_image.sh [hunk 8]
```diff
@@ -18,22 +18,20 @@ if [ "$INPUT_APPENDIX_FILE" ]; then
...
 # Set image name to username/repo_name if not provided
 if [ -z "$INPUT_IMAGE_NAME" ]; then
+    if [[ -z "$INPUT_DOCKER_USERNAME" ]]; then
```
Comment (yuvipanda) [inline create_docker_image.sh:30]: I think this is causing https://github.com/jupyterhub/repo2docker-action/issues/74. If IMAGE_NAME is not set but `NO_PUSH` is set, this is all ok...

**Pull request merged**